### ISSUE TITLE
Add pre-compiled ARM binaries for releases

### ISSUE THREAD

**Issue by zaolin** [concourse/concourse#1379]
Hey guys,

# Feature Request
I wanted to use concourse ci workers on my raspberry pi 3 for doing QA with x86 firmware.
The QA should be really cheap < 100$ so that people can attach their own test stand.
Would it be possible to deliver pre-compiled ARM binaries for releases ?

Best Regards, Zaolin

---

Edit from @taylorsilva April 2022

No plans for official ARM images yet because it's hard to add this to our process. AWS is the only cloud provider with ARM instances and we do everything on GCP currently. Adding a single ARM instance is therefore really hard. This will be more feasible once more cloud providers add ARM instances to their offerings.

Currently this comment, farther down in this issue, is your best option: https://github.com/concourse/concourse/issues/1379#issuecomment-929625288

---

Edit from @taylorsilva January 2025

There's been some progress on this front and I think we can expect some official ARM binaries this year. Stay tuned!

---

## TODO

- [ ] Build ARM version of base resource-types using buildkit qemu
  - [ ] git
  - [ ] s3
  - [ ] registry-image
  - [ ] time
  - [ ] mock
  - [ ] github-release
  - [ ] pool
  - [ ] docker-image
  - [ ] hg
  - [ ] semver
  - [ ] bosh-release
  - [ ] bosh-stemcell
- [ ] Fly
  - [ ] Linux
  - [ ] Mac
- [ ] Update web UI so users can select OS/ARCH combo they want. Maybe include command line instructions directly in the web UI
- [ ] Build ARM Concourse image and binary
  - [ ] Linux
  - [ ] Mac

**Comment by Niraj-Fonseka** [comment]
What is the status on this ?

**Comment by jama-pivotal** [comment]
@Niraj-Fonseka its currently in our icebox under the Operations project. Unfortunately its not a priority for the team right now

**Comment by brownjohnf** [comment]
@jama-pivotal I'm also interested in this, and open to building my own binaries, but I can't find any documentation anywhere about how to build the set of concourse binaries. Could you please point me to documentation?

**Comment by jama-pivotal** [comment]
@brownjohnf we have some getting started with contributing under our quickstart guide for engineer contributors: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md 

Not sure if that has enough specific detail to get you started though...

/cc @vito to see if he has any more resources

**Comment by neumayer** [comment]
I also tried to take a look at that and to be honest I struggled quite a bit.

I looked both at the existing pipelines at https://github.com/concourse/pipelines to run with my own concourse setup and tried to build things locally on my machine (and various combinations thereof). And didn't get very far... I have a local concourse setup I can experiment with, it just seems to duplicate the whole pipeline touches on so many custom things (versioning in s3, all the bosh stuff maybe?) that shouldn't be necessary just to build and yet they seem so integrated.

So any help would really be appreciated. I might be able to help out with various things in that context since we do have a new requirement for building arm stuff here. The concrete thing I need really is a concourse binary built with GOARCH=arm.

**Comment by petrosagg** [comment]
### tl;dr
There are [armv7 and aarch64 builds available](https://github.com/resin-io/concourse-arm/releases/tag/v3.14.1) that work with caveats.

### Building

I had a go on this and I went down a deep rabbit hole but I finally managed to cross compile both an `armv7` and `aarch64` concourse binary. This was a chicken and an egg problem because the normal build process of concourse uses concourse itself and the build process pulls in various prebuilt resources from other build processes.

So I took the path of building everything from scratch using a bash script that bootstraps a concourse binary. The bash script more or less follows what the pipeline does. The idea is that this stage0 concourse binary can then be used on an actual arm machine to run the normal build pipeline and produce the concourse binary.

There are some build issues with both armv7 and aarch64 of which I think I fixed all of them and some of them are already upstreamed:

https://github.com/concourse/baggageclaim/pull/11
https://github.com/cloudfoundry/guardian/pull/118
https://github.com/concourse/time-resource/pull/31

You can find the bootstrap repo here https://github.com/resin-io/concourse-arm. To cross compile it all you need is a linux system with a working Go and Docker installation. You can also find pre-compiled binaries in the Github releases page.

### Running

I haven't yet tested the arm64 version but I believe it will work without issues.

Unfortunately there are still some runtime issues for armv7 builds which are due to a bug of golang's syscall module. If you attempt to run concourse on 32bit hardware with user namespaces enabled you'll get:

```
panic: integer overflow on token 4294967295 while parsing line "         0          0 4294967295"

goroutine 1 [running]:
github.com/concourse/baggageclaim/uidgid.must(0x0, 0xc3050c0, 0x1d7d9650, 0xc3050c0)
	/home/petrosagg/projects/concourse-arm-stage0/workdir/concourse/src/github.com/concourse/baggageclaim/uidgid/max_valid_uid.go:81 +0x40
github.com/concourse/baggageclaim/uidgid.MustGetMaxValidUID(0x0)
	/home/petrosagg/projects/concourse-arm-stage0/workdir/concourse/src/github.com/concourse/baggageclaim/uidgid/max_valid_uid.go:22 +0x40
github.com/concourse/baggageclaim/uidgid.NewPrivilegedMapper(0x1659e01, 0x5)
	/home/petrosagg/projects/concourse-arm-stage0/workdir/concourse/src/github.com/concourse/baggageclaim/uidgid/mapper_linux.go:14 +0x14
github.com/concourse/baggageclaim/baggageclaimcmd.
[...]
```
This is because the entries in `/proc/<pid>/uid_map` [are unsigned 32bit integers](https://github.com/torvalds/linux/blob/v4.17/include/linux/user_namespace.h#L17-L21) but baggageclaim uses golang's syscall module which [defines them as `int`s](https://github.com/golang/go/blob/5c11480631f5654e9e6937ff08c453660138c64d/src/syscall/exec_linux.go#L14-L20).

A potential way forward (fixing go aside) would be to use runc's libraries that has already fixed this issue https://github.com/opencontainers/runc/pull/1819 . It's still the wrong datatype but at least it doesn't overflow.

### Resource types

The official concourse binaries currently ship with the following resource types embedded:

* bosh-deployment
* bosh-io-release
* bosh-io-stemcell
* cf
* docker-image
* git
* github-release
* hg
* pool
* s3
* semver
* time
* tracker

Of those I only cross-compiled:

* docker-image
* git
* s3
* time

This means that the current binaries are not yet able to run the full concourse pipeline and build itself since it uses more resource types but you can definitely use it if your pipelines don't need those resource types. Depending on the resource type, cross compiling could be as simple as [switching the base image](https://github.com/resin-io/concourse-arm/blob/b5218a/patches/arm/git-resource/0001-build-for-arm.patch). I expect bosh to be the most tricky one.

### Upstreaming

Currently the build process changes the projects in a way that breaks normal amd64 builds. So there is work left to be done to have a multi-arch build process that can be upstreamed.

**Comment by vito** [comment]
@petrosagg Thanks for looking into this!

**Comment by neumayer** [comment]
Yeah, thanks a lot. I'll try to set this up myself and will report back on how it goes :-)

**Comment by petrosagg** [comment]
I just pushed a couple of PRs for the int overflow issue

https://github.com/concourse/baggageclaim/pull/12
https://github.com/cloudfoundry/idmapper/pull/3

and published a new arm binary as v3.14.1-rc2 :)

https://github.com/resin-io/concourse-arm/releases/tag/v3.14.1-rc2

It now initialises correctly and I can load the UI. Haven't done further testing or actual builds yet

**Comment by neumayer** [comment]
I managed to successfully run an aarch64 worker (before your update). 

Now I just need to figure out how to bootstrap the aarch64 image used to build another aarch64 image in my pipeline :-)

**Comment by petrosagg** [comment]
@neumayer nice! What do you mean by bootstrapping the aarch64 image? We (resin) build the cross platform base images that I used for this so if you have any questions about how the cross-compilation works or how to run them natively I can help you

**Comment by neumayer** [comment]
The actual task I'm trying to solve is to make concourse build (and publish) docker images for me.
In my normal (x86_64) workflow I use a dind image to build docker images (well build and run inspec tests). Naturally that that one dind image that I usually use is not aarch64. It is built with concourse itself in another pipeline. So I think what I will look at next is how to create multi-arch images in a nice way in concourse. 

This is a bit new to me but I'll keep you posted once I get that working (and especially if I don't get it working :-)). Thanks for the offer, I just need to find the time to read up on the multi arch docker thing a bit first.

**Comment by neumayer** [comment]
Hi!

I have some updates on my multi arch journey :-)

One thing I thought would be great for this whole thing is to build all the images involved multi-arch aware. I.e. the images I want to use in concourse should exist for both amd64 and aarch64 architectures and this is easily accomplished by using alpine based images. Wherever they are built the architecture is properly propagated and the right repos are set up and the packages for the right architecture are installed. I don't know how this works for other base images (I do expect some issues here though, and at some point I think it'll be inevitable to have one Dockerfile per architecture or lots of if statements in the Dockerfiles).

I build a modified dind image that supports multi arch which in turn can be used to build other docker images. It's a bit of a chicken and egg problem, but once the initial image is bootstrapped you're ready to go.

Then I made a concourse pipeline to use this image with to tasks, one for amd64 (i.e. the normal job) building a xxx:amd64 image and one that has a tag specified so the right worker is used (aarch64 is the tag I use for the aarch64 workers) to build a xxx:aarch64 image. There's a third task building a manifest and pushing the manifest for both images (xxx -> xxx:amd64, xxx:aarch64).

I'm quite new to the multi-arch stuff in docker, but it seems to me that the only viable way forward is to add architecture tags to all images anyway (so the manifests can refer to them later).

There's one thing I noticed: when concourse pulls the image for the aarch64 task, i.e. the one that is run on the aarch64 host, it seems to ask for the amd64 image explicitly (via the architecture setting in docker). So my assumption is that the code somewhere has some custom logic that falls back to the amd64 architecture rather than propagating the architecture of the worker in a generic way. I'll double check that if I find the time, it's just a theory so far. When I pull the image without tags from inside the build container the right image is pulled (via the docker binary). The easy workaround for now is to add aarch64 tags to both the task and to the image used in the task.

Ideally this would be integrated in the docker image resource somehow, but I don't see how the manifest stuff can fit there, maybe a docker-manifest resource. If both docker-image resource and docker-manifest resources had the same set of architecture tags for which they can check this might work nicely (the docker-image resource has a multi-arch flag that adds these tags and the manifest resource uses the outputs and builds manifests for all known architectures).
If my assumption from before holds the actual code changes would be minimal (just adjust the architecture to be propagated properly).

I'll try to post updates when I have time to look for the architecture propagation code.

**Comment by neumayer** [comment]
I have a short update on my previous ramblings about the architecture propagation code being buggy. That is not true. It works as intended, I had mistagged my images. It happens a lot with building multi-arch images :-), it seems to be the main challenge to keep things consistent.

Anyway. I've been using this rogue worker for aarch64 for almost three months from the binary and so far it's been very stable and worked well together with the concourse-web instance (4.0.0).

So is there any chance of producing official arm 64 binaries? And is there anything I can do to make that easier (even with my limited understanding of the concourse build process)?

**Comment by arrkiin** [comment]
Hey @neumayer, I followed your journey with great interest. Currently I think of using concourse on my raspberry pi 3. So stumbled over this issue. Can you give use a short description or mini tutorial how you achieved your goal? Perhaps we could manage to create a PR to integrate the changes into the build pipeline.

**Comment by neumayer** [comment]
I just used @petrosagg's binary.

**Comment by vielmetti** [comment]
The question of #arm64 builds of Concourse came up yesterday at Dockercon, I am looking forward to seeing this support from my point of view of the @worksonarm project.

**Comment by Bo0mer** [comment]
Just here to give my +1 for having arm build.

**Comment by neumayer** [comment]
I took another look after concourse 5 came out and was hoping the new build structure would be easier to deal with when it comes to arm. Which was true, creating the binary was super easy, just since the resource images are packaged separately now those of course were missing. And I did not investigate how they are built now. At least that's what I remember it's been a couple of months I last looked at this.

But at the latest when the worker protocol changes I'll have to take another look.

**Comment by cirocosta** [comment]
Hey,

I was working on https://github.com/cirocosta/concourse-arm some time ago, which builds Concourse v5.2.0 with Armv7 and Arm64 support, maybe that'd be interesting for you - it's not well polished yet (more of a proof of concept before bringing a better structured multi-arch support for our builds), but at the moment, it works well (despite the lack of `seccomp` - see the `guardian` fork as a submodule in the repo).

Thanks!

**Comment by analytically** [comment]
Any update on this? Given AWS is pushing Graviton2 as next gen it'd be really handy if Concourse came with arm64 support out of the box.

**Comment by taylorsilva** [comment]
Definitely see this getting higher priority in the future but it's currently not on the road map

**Comment by analytically** [comment]
@taylorsilva @vito given Apple's M1 chip is also arm64, I think it should definitely be added to the road map

**Comment by neumayer** [comment]
In the meantime, I can offer this.

We've been using native arm workers via docker for quite some time now, and containerd on arm has been working well here for months now. I have uploaded my almost self-contained build setup to create a multi platform docker image to run workers with. I say almost self contained because the registry-image (the only resource needed to run build jobs on arm) must be built first, then zipped. The reason for this really is to get to use the alpine image that is used in the registry-image resource, the arm version of that.

https://github.com/neumayer/concourse-arm-worker

It does not cross-compile, so you need to have a working arm machine with docker to build (so it uses the correct arm versions of the alpine images used).
It is not 100 percent up to date, but uses the git sources for concourse v6.5.1 (roughly).
The way I build it is with a concourse task that runs on the existing arm worker with multiarch docker in docker build (not shared, but it basically calls the build script in a dind-arm container).

I wanted to update to the newest release soonish, if I remember I can try updating here and on github (I remember reading some new guardian versions and parameter changes from the release pages, but should be fine to integrate with this).

**Comment by cirocosta** [comment]
> Given AWS is pushing Graviton2 as next gen it'd be really handy if Concourse came with arm64 support out of the box.

and yeah, the numbers tell a quite nice story for Graviton2 when it comes to performance/$:

https://www.phoronix.com/scan.php?page=article&item=amazon-graviton2-benchmarks&num=7

> _"To little surprise, the M6g instances were generally offering the best performance-per-dollar."_

**Comment by vito** [comment]
Our team shrank by quite a bit recently, so I don't really know when we'll have the bandwidth for this, but I'll optimistically put it at the bottom of the (pretty shallow) backlog so we don't forget about it since it definitely makes sense to at least have this on the radar.

Realistically, this seems like it might involve a lot of pipeline plumbing and infrastructure support, so it's hard to square something like this with the other high-priority items on the roadmap (v10, native K8s runtime) given that we can only really do 3 things in parallel at the moment.

If anyone is interested in attempting to move this forward from the outside, e.g. by forking our [pipelines](https://github.com/concourse/ci/tree/master/pipelines) and submitting PRs, we'd be happy to help.

/cc @scottietremendous @matthewpereira

**Comment by martin-g** [comment]
One more +1 for official aarch64 binaries!

**Comment by emiliofernandes** [comment]
Adding support for ARM64 would be awesome!
I hope it would be done sooner!
Great work @concourse/team !

**Comment by martin-g** [comment]
I've written a blog article how to run a worker node on ARM64 based on https://github.com/neumayer/concourse-arm-worker - https://martin-grigorov.medium.com/concourseci-on-arm64-229833883ca9
Thanks to @neumayer and @taylorsilva for their help!

**Comment by marco-m-pix4d** [comment]
(optimist mood)
Would it be fair to say that, for Apple M1, the situation would be way simpler?
Since the Concourse worker on M1 would have macOS as OS, it would not have to care about being able to run containers and Concourse resources :-)

**Comment by vito** [comment]
@marco-m-pix4d Yep - building that really shouldn't involve anything more than just `go build ./cmd/concourse`.

**Comment by dixie-flatline-rom** [comment]
Never mind, aarch64 still loves you!

**Comment by taylorsilva** [comment]
Someone started some work in this PR but it's stalled: https://github.com/concourse/ci/pull/386
If someone wants to pick it up the team can provide guidance.

**Comment by robinhuiser** [comment]
I've created a [repository](https://github.com/robinhuiser/concourse-arm64) to build `arm64` images.

This repository helps you build both the web and worker `arm64` components for Concourse CI - prebuilt Docker images can be found on Docker Hub [rdclda/concourse](https://hub.docker.com/repository/docker/rdclda/concourse).

## Bundled resources

| Concourse | [git](https://github.com/concourse/git-resource) | [github-release](https://github.com/concourse/github-release-resource) | [registry-image](https://github.com/concourse/registry-image-resource) | [semver](https://github.com/concourse/semver-resource) | [time](https://github.com/concourse/time-resource) | [mock](https://github.com/concourse/mock-resource) | [s3](https://github.com/concourse/s3-resource) | [slack-alert](https://github.com/arbourd/concourse-slack-alert-resource) |
|--- |--- |--- |--- |--- |--- |--- |--- |--- |
| v7.1.0 | v1.12.0 | v1.5.2 | v1.2.0 | v1.3.0 | v1.5.0 | v0.11.1 | v1.1.1 | v0.15.0 |
| v7.2.0 | v1.12.1 | v1.5.2 | v1.2.1 | v1.3.0 | v1.6.0 | v0.11.1 | v1.1.1 | v0.15.0 |
| v7.3.2 | v1.14.0 | v1.6.1 | v1.3.0 | v1.3.1 | v1.6.0 | v0.11.2 | v1.1.1 | v0.15.0 |
| v7.4.0 | v1.14.0 | v1.6.4 | v1.4.0 | v1.3.4 | v1.6.1 | v0.12.2 | v1.1.2 | v0.15.0 |
| v7.5.0 | v1.14.4 | v1.6.4 | v1.4.1 | v1.3.4 | v1.6.2 | v0.12.3 | v1.1.3 | v0.15.0 |

**Comment by sahmeder** [comment]
Hey there @robinhuiser . Sorry if this is a dumb question, but I've tried to run that docker image and I keep getting the error message "error: Please specify one command of: generate-key, land-worker, migrate, quickstart, retire-worker, web or worker". Is it even possible to run Concourse on M1 AArch64/Arm64?

**Comment by robinhuiser** [comment]
Hi @sahmeder - I have tested the arm64 images on Raspberry Pi only, but I would be happy to help out. Could you please share your deployment (docker-compose, Kubernetes manifests) so I can be of assistance?

**Comment by sahmeder** [comment]
Hey @robinhuiser Thanks so much for responding :)

What I did was run the rdclda/concourse docker image that you [created here](https://github.com/robinhuiser/concourse-arm64/blob/main/docker-compose.yaml). I was getting the errror message that my iptables were failing and someone created a github issue here where using the rdclda/concourse image [fixed that issue for them](https://githubmemory.com/repo/docker/for-mac/issues/5547). 

I tried to run a container and use your image which is when I got the "error: Please specify one command of: generate-key, land-worker, migrate, quickstart, retire-worker, web or worker". I did some more digging and it looks like this message happens when the docker run privilege isn't set to true, but when I checked the docker-compose file, it's showing that it's set to privilege. I've been scratching my head trying to get this figured out but I'm at a lost so far lol.

**Comment by robinhuiser** [comment]
Hi @sahmeder - I will replay your issue later tonight, need access to an M1 MacBook or Raspberry Pi which I do not have now.

**Comment by sahmeder** [comment]
@robinhuiser  No rush :)

**Comment by sahmeder** [comment]
I wanted to add that I tried installing it via a Ubuntu CLI through multipass and got the same results as well.

**Comment by robinhuiser** [comment]
@sahmeder  - did you try to use the provided docker-compose.yaml file in the Docker registry linked Git repo?

~~~yaml
version: '3'

services:
  concourse-db:
    image: postgres
    environment:
      POSTGRES_DB: concourse
      POSTGRES_PASSWORD: concourse_pass
      POSTGRES_USER: concourse_user
      PGDATA: /database

  concourse:
    image: rdclda/concourse:7.5.0
    command: quickstart
    privileged: true
    depends_on: [concourse-db]
    ports: ["8080:8080"]
    environment:
      CONCOURSE_POSTGRES_HOST: concourse-db
      CONCOURSE_POSTGRES_USER: concourse_user
      CONCOURSE_POSTGRES_PASSWORD: concourse_pass
      CONCOURSE_POSTGRES_DATABASE: concourse
      # replace this with your external IP address
      CONCOURSE_EXTERNAL_URL: http://10.0.19.18:8080
      CONCOURSE_ADD_LOCAL_USER: test:test
      CONCOURSE_MAIN_TEAM_LOCAL_USER: test
      # instead of relying on the default "detect"
      CONCOURSE_WORKER_BAGGAGECLAIM_DRIVER: overlay
      CONCOURSE_CLIENT_SECRET: Y29uY291cnNlLXdlYgo=
      CONCOURSE_TSA_CLIENT_SECRET: Y29uY291cnNlLXdvcmtlcgo=
      CONCOURSE_X_FRAME_OPTIONS: allow
      CONCOURSE_CONTENT_SECURITY_POLICY: "*"
      CONCOURSE_CLUSTER_NAME: arm64
      CONCOURSE_WORKER_CONTAINERD_DNS_SERVER: "8.8.8.8"
      CONCOURSE_WORKER_RUNTIME: "containerd"
~~~

**Comment by sahmeder** [comment]
Hey @robinhuiser ! I was able to get it up and running. So what I was doing incorrectly was that instead of zipping the concourse-arm64 folder and running docker-compose up within that folder locally, what I was doing was pulling the docker image from the docker registry. A little new to docker and containerization so please excuse my ignorance :)

**Comment by natto1784** [comment]
Hello @robinhuiser, have you ever experienced something like https://github.com/concourse/concourse-docker/issues/79 ?

**Comment by DonBower** [comment]
I would love to run concourse with ansible on raspberry pi’s k3s or native

**Comment by robinhuiser** [comment]
@DonBower  - please check out the prebuilt Docker images for Raspberry Pi - they can be found on Docker Hub [rdclda/concourse](https://hub.docker.com/repository/docker/rdclda/concourse).

**Comment by sfxworks** [comment]
Daily driving a rpi4 at the office. Would definitely like this.

**Comment by ccwienk** [comment]
+1 for arm64-support. Maybe it is worth mentioning that GCP by now also seems to offer [arm64-VMs](https://cloud.google.com/compute/docs/instances/arm-on-compute#tau_t2a_machine_series).

I could build and deploy a multi-platform (x86_64 and arm64) instance of concourse using @robinhuiser 's build-scripts basically "out-of-the-box", so it seems there is not that much missing (except, maybe, some adjustments in upstream helm-chart). I figure, (native) multiarch support might become more relevant w/ arm gaining popularity (after all, this is why I setup multi-platform concourse).

**Comment by rfpludwick** [comment]
+1 for arm64; I'd like to use this in my home lab ARM k8s cluster.

**Comment by DonBower** [comment]
+1 apple silicon MacBooks
Especially for fly...

**Comment by natto1784** [comment]
bump
btw i already use fly on aarch64 from nixpkgs like i did earlier in feb, so compiling it on arm64 should not be an issue.

**Comment by xtremerui** [comment]
Hi all, thank you for your interest and patience on this topic. 

With the [multi-arch feature](https://github.com/concourse/oci-build-task/pull/93) now available in oci-build-task, after running a spike on Concourse CI to build concourse/dev with multi-arch, the path is clear for us to have the ARM release: that we need the following dependencies to be available in ARM.

The dependencies:

1. [concourse/golang-builder](https://hub.docker.com/r/concourse/golang-builder) (available in ARM already)
2. concourse/resource-types-(alpine/ubuntu)-image in where all Concourse built-in resource types' binary have to be ARM based.
3. fly binaries
4. [containerd](https://github.com/containerd/containerd/releases), [CNI](https://github.com/containernetworking/plugins/releases), [runc](https://github.com/opencontainers/runc/releases) and [gdn](https://github.com/cloudfoundry/garden-runc-release/releases) for worker runtime
5. And finally [concourse/dev](https://hub.docker.com/r/concourse/dev). Once this one is ready, Concourse ARM release will be just a rebuild from it with harden dependencies. In fact, during the spike, this image is available in ARM already(with resource-types and fly binaries in AMD that will certainly break any check or build though). 

The blocker:

-  ATM, [gdn release](https://github.com/cloudfoundry/garden-runc-release/releases) is only available in AMD. For building a concourse/dev image, we could fetch the source code and [compile it](https://github.com/concourse/ci/blob/bd7094f54fdb73a5ba824e2d34632e3bcbdb4eed/dockerfiles/dev-multi-arch/Dockerfile#L36-L39) in runtime. But for a formal Conocurse ARM release we will need to pin it to a specific version. So it depends on when the upstream could have a `gdn` release for ARM.
- Performance. `docker buildx` emulates arm/v7 in a AMD system to cross build the target for ARM. During this process, it takes about 10x times longer to compile above dependency libraries, which adds huge burden to our CI and often leads to unresponsive builds. For example, in one build of registry-image resource, the [build](https://ci.concourse-ci.org/teams/main/pipelines/resource/jobs/build-alpine/builds/106?vars.type=%22registry-image%22) hangs forever probably due to exhausted container resources. In the build log that one finished successfully, you can tell the differences of time using for building resource binaries for `linux/amd64` and `linux/arm64`.
```
#18 [builder 10/10] RUN set -e; for pkg in $(go list ./...); do 		go test -o "/tests/$(basename $pkg).test" -c $pkg; 	done
#18 13.56 ?   	github.com/concourse/registry-image-resource/cmd/check	[no test files]
#18 14.01 ?   	github.com/concourse/registry-image-resource/cmd/in	[no test files]
#18 14.41 ?   	github.com/concourse/registry-image-resource/cmd/out	[no test files]
#18 14.87 ?   	github.com/concourse/registry-image-resource/commands	[no test files]
#18 DONE 15.1s
```
```
#36 [linux/arm64 builder 10/10] RUN set -e; for pkg in $(go list ./...); do 		go test -o "/tests/$(basename $pkg).test" -c $pkg; 	done
#36 286.2 ?   	github.com/concourse/registry-image-resource/cmd/check	[no test files]
#36 291.7 ?   	github.com/concourse/registry-image-resource/cmd/in	[no test files]
#36 297.1 ?   	github.com/concourse/registry-image-resource/cmd/out	[no test files]
#36 302.7 ?   	github.com/concourse/registry-image-resource/commands	[no test files]
#36 DONE 302.9s
```
The plan:

- we will submit an issue to upstream gdn release to ask for a release in ARM. Or maybe in one day we don't need to use `guardian` runtime (it has problem to run in ubuntu jammy too) ...? 
- we might probably separate the build task of AMD and ARM into different workers (so ARM build task can run in a native ARM vm that available by GCP) using a solution similar to this [gist](https://gist.github.com/jruiz94/f4b7bbb67e0ff7c71e9c582c7673af0c), if the performance issue is not improved overtime. That might require registry-image-resource/docker-image-resource to support push/pull by cache layer. And also, we will have to manually build a Concourse ARM release to run the ARM based worker (which comes first chicken or eggs :D ).

Please let us know your thoughts or ideas. Thank you!

**Comment by mvdkleijn** [comment]
@xtremerui I see in https://github.com/cloudfoundry/garden-runc-release/releases that gdn now has arm builds since 1.22.9

That removes at least one blocker.

Is your second point of the above mentioned plan still valid? In other words, is performance still a concern or would it be possible to start making arm releases of Concourse? I'd really like to get this off the ground as it were..

**Comment by xtremerui** [comment]
@mvdkleijn yes it is now happening, slowly... The [concourse/dev](https://hub.docker.com/r/concourse/dev/tags) image supports ARM platform already. So running Concourse docker compose on a M1 machine should work flawlessly. I can see the compilation time of Concourse binary in the image takes less time significantly. That might allow our CI to adapt ARM faster i.e. without separating the build task.

As the ARM support propagating through our CI, eventually we should be able to release Concourse in ARM. However there is no timeline for that yet.

**Comment by KedXP** [comment]
Tried dev:latest.
Web running on amd64, worker running on arm64 (orange pi 5)
Worker running with env variables
```
CONCOURSE_BAGGAGECLAIM_DRIVER="naive"
CONCOURSE_RUNTIME="houdini"
```
on hello world job with arm64v8/alpine image get error 
`fork/exec /opt/resource/check: exec format error`

Full worker env
```
CONCOURSE_WORK_DIR=/opt/concourse/worker
CONCOURSE_TSA_HOST=my-web-pc:2222
CONCOURSE_TSA_PUBLIC_KEY=/keys/tsa_host_key.pub
CONCOURSE_TSA_WORKER_PRIVATE_KEY=/keys/worker_key
CONCOURSE_TAG="arm64"
CONCOURSE_CONTAINERD_DNS_SERVER="8.8.8.8"
CONCOURSE_BAGGAGECLAIM_DRIVER="naive"
CONCOURSE_RUNTIME="houdini"
```

docker-compose for worker
```
version: '3'

services:
    worker:
      image: concourse/dev:latest
      # image: rdclda/concourse:7.9.1
      # image: concourse-arm-worker:latest
      command:
        - worker
      privileged: true
      volumes:
        - ./keys:/keys
      env_file:
        - ./configs/worker.env
```
log for job
```
concourse-worker-1  | {"timestamp":"2023-09-15T01:04:48.395937795Z","level":"info","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.get-volume.volume-not-found","data":{"session":"4.1.57.1","volume":"b17c31ae-2af5-4ca2-402e-e65b1593249e"}}
concourse-worker-1  | {"timestamp":"2023-09-15T01:04:48.396505367Z","level":"info","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.volume-not-found","data":{"session":"4.1.57","volume":"b17c31ae-2af5-4ca2-402e-e65b1593249e"}}
concourse-worker-1  | {"timestamp":"2023-09-15T01:04:49.795298253Z","level":"info","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.get-volume.volume-not-found","data":{"session":"4.1.63.1","volume":"736b34ce-bbe2-44ab-4dbd-64e1f39d9762"}}
concourse-worker-1  | {"timestamp":"2023-09-15T01:04:49.795384877Z","level":"info","source":"baggageclaim","message":"baggageclaim.api.volume-server.get-volume.volume-not-found","data":{"session":"4.1.63","volume":"736b34ce-bbe2-44ab-4dbd-64e1f39d9762"}}
concourse-worker-1  | {"timestamp":"2023-09-15T01:04:49.838992561Z","level":"info","source":"worker","message":"worker.garden.garden-server.create.created","data":{"request":{"Handle":"4b2ccaff-e231-49ed-4b36-fd93b9eb539c","GraceTime":0,"RootFSPath":"raw:///opt/concourse/worker/volumes/live/b17c31ae-2af5-4ca2-402e-e65b1593249e/volume","BindMounts":[{"src_path":"/opt/concourse/worker/volumes/live/736b34ce-bbe2-44ab-4dbd-64e1f39d9762/volume","dst_path":"/scratch","mode":1}],"Network":"","Privileged":false,"Limits":{"bandwidth_limits":{},"cpu_limits":{},"disk_limits":{},"memory_limits":{},"pid_limits":{}}},"session":"1.2.41"}}
concourse-worker-1  | {"timestamp":"2023-09-15T01:04:49.877225810Z","level":"info","source":"worker","message":"worker.garden.garden-server.get-properties.got-properties","data":{"handle":"4b2ccaff-e231-49ed-4b36-fd93b9eb539c","session":"1.2.42"}}
concourse-worker-1  | {"timestamp":"2023-09-15T01:04:49.894437014Z","level":"error","source":"worker","message":"worker.garden.garden-server.run.failed","data":{"error":"fork/exec /opt/resource/check: exec format error","handle":"4b2ccaff-e231-49ed-4b36-fd93b9eb539c","session":"1.2.43"}}
```
pipeline
```
jobs:
- name: hello-world-job
  plan:
  - task: echo-print
    tags: [arm64]
    config:
      platform: linux
      image_resource:
        type: registry-image
        source:
          repository: arm64v8/alpine
          tag: latest # images are pulled from docker hub by default
          platform:
            architecture: arm64
            os: linux
      run:
        path: /bin/echo
        args: ["123123123"]
```

**Comment by NorseGaud** [comment]
Hello, what's the status on supporting arm64 for darwin?

**Comment by xtremerui** [comment]
@NorseGaud there is really no timeline yet.

**Comment by HariSekhon** [comment]
+1 for arm64 for new Macs please

I have [scripts](https://github.com/HariSekhon/DevOps-Bash-tools) which auto-determine the latest GitHub release, OS + architecture and download binaries, hit this on my new Mac:

```
2024-03-03 02:17:17  Downloading: https://github.com/concourse/concourse/releases/download/v7.11.2/fly-7.11.2-darwin-arm64.tgz
https://github.com/concourse/concourse/releases/download/v7.11.2/fly-7.11.2-darwin-arm64.tgz:
2024-03-03 02:17:17 ERROR 404: Not Found.
```

**Comment by taylorsilva** [comment]
Succeeded by #9182
